### PR TITLE
Async dispose

### PR DESCRIPTION
**Pull request by HEskandari**
- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

**PR Title**
Implements IAsyncDisposable in WebApplicationFactory

**PR Description**
Properly implements `IAsyncDisposable` interface on `WebApplicationFactory`. Also includes a functional test.

Addresses #29455

### CODE REVIEW

**Comment by HEskandari** [comment]
@javiercn here's the PR that should address the concerns raised on the older PR [here](https://github.com/dotnet/aspnetcore/pull/29456/).

**Comment by ShreyasJejurkar** [comment]
Thanks for taking care! Appreciated!

**Comment by HEskandari** [comment]
@javiercn Rebased on the main branch and the build is now green. It is good to go as far as I'm concerned.

File: src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs [hunk 1]
```diff
@@ -512,22 +513,53 @@ protected virtual void Dispose(bool disposing)
...
-                _server?.Dispose();
-                _host?.StopAsync().Wait();
+            foreach (var client in _clients)
+            {
+                client.Dispose();
+            }
+
+            foreach (var factory in _derivedFactories)
+            {
+                await factory.DisposeAsync().ConfigureAwait(false);
```
Comment (javiercn) [inline src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs:549]: This should be casted to `IAsyncDisposable` to allow for interface re-implementation on derived classes.

File: src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs [hunk 2]
```diff
@@ -512,22 +513,53 @@ protected virtual void Dispose(bool disposing)
...
 
-                foreach (var factory in _derivedFactories)
-                {
-                    factory.Dispose();
-                }
+            _disposed = true;
+        }
+
+        /// <inheritdoc />
+        public async ValueTask DisposeAsync()
```
Comment (javiercn) [inline src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs:530]: We want to make this virtual to allow people disposing their own resources in derived classes if needed.

File: src/Mvc/Mvc.Testing/src/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -49,5 +49,6 @@ virtual Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory<TEntryPoint>.Crea
 virtual Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory<TEntryPoint>.CreateHostBuilder() -> Microsoft.Extensions.Hosting.IHostBuilder?
 virtual Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory<TEntryPoint>.CreateServer(Microsoft.AspNetCore.Hosting.IWebHostBuilder! builder) -> Microsoft.AspNetCore.TestHost.TestServer!
 virtual Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory<TEntryPoint>.CreateWebHostBuilder() -> Microsoft.AspNetCore.Hosting.IWebHostBuilder!
+virtual Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory<TEntryPoint>.DisposeAsyncCore() -> System.Threading.Tasks.ValueTask
```
Comment (javiercn) [inline src/Mvc/Mvc.Testing/src/PublicAPI.Unshipped.txt:52]: We don't have this method anywhere (and we shouldn't have it)

Comment (javiercn) [reply src/Mvc/Mvc.Testing/src/PublicAPI.Unshipped.txt:52]: I think the baseline needs to be updated

Comment (HEskandari) [reply src/Mvc/Mvc.Testing/src/PublicAPI.Unshipped.txt]: @javiercn tried removing it, but the build fails [like so](https://github.com/dotnet/aspnetcore/pull/32645/checks?check_run_id=2684163922).

Comment (HEskandari) [reply src/Mvc/Mvc.Testing/src/PublicAPI.Unshipped.txt]: @javiercn the last changes pass all the builds.

File: src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs (hunk 2 above)
Comment (HEskandari) [reply src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs]: done

File: src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs (hunk 1 above)
Comment (HEskandari) [reply src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs]: done

**Comment by HEskandari** [comment]
@javiercn anything else needed to be done here?

**Review by javiercn — APPROVED** [review]
Looks great!

**Comment by javiercn** [comment]
Thanks for the contribution!

**Pull request merged**